### PR TITLE
events: propagate hubID over the events

### PR DESCRIPTION
**Pull request by jkralik**
Assign a `hubID` to every generated event using the identity-store and resource-aggregate mechanisms. This enables the differentiation of hubs in cases where your application spans multiple independent hub deployments.

### CODE REVIEW

File: grpc-gateway/pb/doc.html [hunk 1]
```diff
@@ -3694,6 +3694,13 @@ <h3 id="resourceaggregate.pb.EventMetadata">EventMetadata</h3>
                   <td><p>sequence number within the same connection_id; the ResourceChanged event uses the value to skip old events, other event types might not fill the value </p></td>
                 </tr>
               
+                <tr>
+                  <td>hub_id</td>
+                  <td><a href="#string">string</a></td>
+                  <td></td>
+                  <td><p> </p></td>
```
Comment (ondrejtomcik) [inline grpc-gateway/pb/doc.html]: Missing description of the property

File: grpc-gateway/pb/service.swagger.json [hunk 2]
```diff
@@ -1510,6 +1510,9 @@
           "type": "string",
           "format": "uint64",
           "title": "sequence number within the same connection_id; the ResourceChanged event uses the value to skip old events, other event types might not fill the value"
+        },
+        "hubId": {
+          "type": "string"
```
Comment (ondrejtomcik) [inline grpc-gateway/pb/service.swagger.json]: Missing title and format

File: identity-store/config.yaml [hunk 3]
```diff
@@ -55,6 +55,7 @@ clients:
         certFile: "/secrets/public/cert.crt"
         useSystemCAPool: false
   storage:
+    hubID: ""
```
Comment (ondrejtomcik) [inline identity-store/config.yaml]: Why is it part of storage and not part of the root service configuration?

File: identity-store/persistence/persist.go [hunk 4]
```diff
@@ -4,6 +4,7 @@ package persistence
 type AuthorizedDevice struct {
 	DeviceID string `db:"device_id"`
 	Owner    string `db:"owner"`
+	HubID    string `db:"hub_id"`
```
Comment (ondrejtomcik) [inline identity-store/persistence/persist.go]: I am not aware of the plan to reuse one identity store service for multiple instances of plgd hub. Why do you need to persist the value?
What if the hub id changes? How do you reconcile the database? Should you still see devices - what about querying?

File: identity-store/service/config.go [hunk 5]
```diff
@@ -74,12 +74,16 @@ func (c *EventBusConfig) Validate() error {
...
 type StorageConfig struct {
 	MongoDB mongodb.Config `yaml:"mongoDB" json:"mongoDb"`
+	HubID   string         `yaml:"hubID" json:"hubID"`
 }
 
 func (c *StorageConfig) Validate() error {
 	if err := c.MongoDB.Validate(); err != nil {
 		return fmt.Errorf("mongoDB.%w", err)
 	}
+	if c.HubID == "" {
```
Comment (ondrejtomcik) [inline identity-store/service/config.go]: We need to validate that the hub ID is a valid uuid.

File: resource-aggregate/config.yaml [hunk 6]
```diff
@@ -62,6 +62,7 @@ clients:
     snapshotThreshold: 16
     # limits number of try to store event
     occMaxRetry: 8
+    hubID: ""
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: Please unify the location of this configuration option. In the identity store, you have it in the storage. Here you have it under the event store. Please consider moving it to the root.

File: resource-aggregate/service/config.go [hunk 7]
```diff
@@ -81,6 +82,9 @@ func (c *EventStoreConfig) Validate() error {
 	if c.ConcurrencyExceptionMaxRetry <= 0 {
 		return fmt.Errorf("occMaxRetry('%v')", c.ConcurrencyExceptionMaxRetry)
 	}
+	if c.HubID == "" {
```
Comment (ondrejtomcik) [inline resource-aggregate/service/config.go]: Don't forget to validate the format.

File: resource-aggregate/service/aggregate.go [hunk 8]
```diff
@@ -17,25 +17,27 @@ type LogPublishErrFunc func(err error)
...
 }
 
 func DeviceMetadataFactoryModel(context.Context) (cqrsAggregate.AggregateModel, error) {
-	return raEvents.NewDeviceMetadataSnapshotTaken(), nil
+	return events.NewDeviceMetadataSnapshotTaken(), nil
 }
 
 // NewAggregate creates new resource aggreate - it must be created for every run command.
-func NewAggregate(resourceID *commands.ResourceId, snapshotThreshold int, eventstore EventStore, factoryModel cqrsAggregate.FactoryModelFunc, retry cqrsAggregate.RetryFunc) (*Aggregate, error) {
+func NewAggregate(resourceID *commands.ResourceId, snapshotThreshold int, hubID string, eventstore EventStore, factoryModel cqrsAggregate.FactoryModelFunc, retry cqrsAggregate.RetryFunc) (*Aggregate, error) {
```
Comment (ondrejtomcik) [inline resource-aggregate/service/aggregate.go:37]: When creating an aggregate, you always pass the hubID from the configuration. When you then load events, you do not care if it is the matching one or not. And when handling command and publishing the event, you always take the hub id from the context -> aggregate -> configuration, while in identity store you take the one stored into the db. You use two different approaches. Please come up with a consistent solution - e.g. updating the identity store.

**Review by ondrejtomcik — CHANGES_REQUESTED** [review]

File: grpc-gateway/pb/doc.html (hunk 1 above)
Comment (jkralik) [reply grpc-gateway/pb/doc.html]: This is generated doc. I don't want to spend time with the generator.

File: identity-store/config.yaml (hunk 3 above)
Comment (jkralik) [reply identity-store/config.yaml]: moved to root

File: grpc-gateway/pb/service.swagger.json (hunk 2 above)
Comment (jkralik) [reply grpc-gateway/pb/service.swagger.json]: fixed

File: identity-store/persistence/persist.go (hunk 4 above)
Comment (jkralik) [reply identity-store/persistence/persist.go]: removed from persistent storage

File: identity-store/service/config.go (hunk 5 above)
Comment (jkralik) [reply identity-store/service/config.go]: fixed

File: resource-aggregate/config.yaml (hunk 6 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: moved to root

File: resource-aggregate/service/config.go (hunk 7 above)
Comment (jkralik) [reply resource-aggregate/service/config.go]: fixed

File: resource-aggregate/service/aggregate.go (hunk 8 above)
Comment (jkralik) [reply resource-aggregate/service/aggregate.go:37]: Fixed -> removed from IS storage

File: charts/plgd-hub/README.md [hunk 9]
```diff
@@ -85,6 +85,7 @@ global:
 | certificateauthority.extraVolumes | string | `nil` | Optional extra volumes |
 | certificateauthority.fullnameOverride | string | `nil` | Full name to override |
 | certificateauthority.httpPort | int | `9101` |  |
+| certificateauthority.hubId | string | `nil` | Hub ID. Can be override via global.hubId or coapgateway.hubId |
```
Comment (Danielius1922) [inline charts/plgd-hub/README.md]: can be overriden

same typo on lines 422 and 570

File: grpc-gateway/pb/devices.proto [hunk 10]
```diff
@@ -112,13 +113,13 @@ message SubscribeToEvents {
 message Event {
   message DeviceRegistered {
     repeated string device_ids = 1;
-
+    identitystore.pb.EventMetadata event_metadata = 5; // provides metadata of event
```
Comment (Danielius1922) [inline grpc-gateway/pb/devices.proto]: why 5 and not 2? are 2-4 reserved?

Comment (jkralik) [reply grpc-gateway/pb/devices.proto]: This is a issue ... `DeviceRegistered/DeviceDeregistered` I want replace them by events from identity store. But we need to introduce new type ...

File: charts/plgd-hub/README.md (hunk 9 above)
Comment (jkralik) [reply charts/plgd-hub/README.md]: fixed

**Review by Danielius1922 — APPROVED** [review]

File: charts/plgd-hub/values.yaml [hunk 11]
```diff
@@ -432,6 +432,8 @@ resourcedirectory:
     mountPath: /config
   # -- Service and POD port
   port: 9100
+  # -- Hub ID. Can be overriden via global.hubId
```
Comment (ondrejtomcik) [inline charts/plgd-hub/values.yaml]: Can be overriden or this can override the global.hubId ?

**Review by ondrejtomcik — APPROVED** [review]

File: charts/plgd-hub/values.yaml (hunk 11 above)
Comment (jkralik) [reply charts/plgd-hub/values.yaml]: fixed

**Pull request merged**